subprojects/core/src/integTest/groovy/org/gradle/api/tasks/AbstractDirectorySensitivityIntegrationSpec.groovy

```diff
@@ -0,0 +1,535 @@
+    }
+
+    @Unroll
+    def "empty directories are ignored when specified (#api, #pathSensitivity)"() {
+        createTaskWithSensitivity(DirectorySensitivity.IGNORE_DIRECTORIES, api, pathSensitivity)
+        buildFile << """
+            taskWithInputs {
+                sources.from(project.files("foo", "bar"))
+                outputFile = project.file("\${buildDir}/output.txt")
+            }
+        """
+        file('foo').mkdir()
+        file('foo/a').createFile()
+        file('foo/b').createFile()
+        file('bar').mkdir()
+        file('bar/a').createFile()
+
+        when:
+        execute("taskWithInputs")
+
+        then:
+        executedAndNotSkipped(":taskWithInputs")
+
+        when:
+        cleanWorkspace()
+        execute("taskWithInputs")
+
+        then:
+        reused(":taskWithInputs")
+
+        when:
+        cleanWorkspace()
+        file('foo/c').mkdir()
+        file('foo/c/1').mkdir()
+        file('foo/c/2').mkdir()
+        execute("taskWithInputs")
+
+        then:
+        reused(":taskWithInputs")
+
+        where:
+        [api, pathSensitivity] << [API.values(), [PathSensitivity.RELATIVE, PathSensitivity.ABSOLUTE, PathSensitivity.NAME_ONLY]].combinations()
+    }
+
+    @Unroll
+    def "Non-empty directories are tracked when empty directories are ignored (#api, #pathSensitivity)"() {
+        createTaskWithSensitivity(DirectorySensitivity.IGNORE_DIRECTORIES, api, pathSensitivity)
+        buildFile << """
+            taskWithInputs {
+                sources.from(project.files("foo", "bar"))
+                outputFile = project.file("\${buildDir}/output.txt")
+            }
+        """
+        file('foo').mkdir()
+        file('foo/a').mkdir()
+        file('foo/a/b').createFile()
+        file('bar').mkdir()
+        file('bar/a').createFile()
+
+        when:
+        execute("taskWithInputs")
+
+        then:
+        executedAndNotSkipped(":taskWithInputs")
+
+        when:
+        cleanWorkspace()
+        execute("taskWithInputs")
+
+        then:
+        reused(":taskWithInputs")
+
+        when:
+        cleanWorkspace()
+        file('foo/a/b') << "foo"
+        execute("taskWithInputs")
+
+        then:
+        executedAndNotSkipped(":taskWithInputs")
+
+        when:
+        cleanWorkspace()
+        file('foo/a').renameTo(file('foo/c'))
+        execute("taskWithInputs")
+
+        then:
+        executedUnlessNameOnly(":taskWithInputs", pathSensitivity)
+
+        when:
+        cleanWorkspace()
+        file('foo').deleteDir()
+        execute("taskWithInputs")
+
+        then:
+        executedAndNotSkipped(":taskWithInputs")
+
+        where:
+        [api, pathSensitivity] << [API.values(), [PathSensitivity.RELATIVE, PathSensitivity.ABSOLUTE, PathSensitivity.NAME_ONLY]].combinations()
+    }
+
+    @ToBeFixedForConfigurationCache(because = "doesn't like FileCollection in transform parameters")
```

Comment:
That is very strange.  An abstract getter for `ConfigurableFileCollection` should work perfectly well for transform parameters.